### PR TITLE
crimson: fix split_pin_left assert_aligned

### PR DESCRIPTION
**Pull request by liu-chunmei**
hint passed to lba alloc_extent may not be block size aligned
in a certain case, this hint insert lba tree as key, so make
laddr in lba tree is not aligned.

Signed-off-by: chunmei-liu <chunmei.liu@intel.com>

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Review by cyx1231st — APPROVED** [review]

**Comment by liu-chunmei** [comment]
jenkins retest make check.

**Comment by liu-chunmei** [comment]
jenkins retest make check

**Comment by athanatos** [comment]
I'd rather fix the caller, I can't think of any cases where it would actually be desirable to pass in an unaligned hint.  Which caller was it?

**Comment by liu-chunmei** [comment]
> I'd rather fix the caller, I can't think of any cases where it would actually be desirable to pass in an unaligned hint. Which caller was it?
onode->get_metadat_hint() I think.  see below:
  laddr_t get_metadata_hint() const {
    assert(default_metadata_offset);
    assert(default_metadata_range);
    return get_hint() + default_metadata_offset +
      ((uint32_t)std::rand() % default_metadata_range);
  }

I can't get block size in onode, I will aligne it in seastore.cc at each place  when call onode->get_meta_hint(). if you think so.

**Comment by athanatos** [comment]
Add an alignment argument to get_meta_hint().

**Comment by liu-chunmei** [comment]
> Add an alignment argument to get_meta_hint().

sure, will do.

**Comment by liu-chunmei** [comment]
@athanatos please review again.

**Comment by liu-chunmei** [comment]
jenkins retest Ceph API please.

File: src/crimson/os/seastore/onode.h [hunk 1]
```diff
@@ -66,11 +66,12 @@ class Onode : public boost::intrusive_ref_counter<
   virtual onode_layout_t &get_mutable_layout(Transaction &t) = 0;
   virtual ~Onode() = default;
 
-  laddr_t get_metadata_hint() const {
+  laddr_t get_metadata_hint(uint64_t block_size) const {
     assert(default_metadata_offset);
     assert(default_metadata_range);
-    return get_hint() + default_metadata_offset +
+    laddr_t hint = get_hint() + default_metadata_offset +
       ((uint32_t)std::rand() % default_metadata_range);
```
Comment (athanatos) [inline src/crimson/os/seastore/onode.h:73]:
```
uint64_t range_blocks = default_metadata_range / block_size;
laddr_t hint = get_hint() + default_metadata_offset +
  (((uint32_t)std::rand() % range_blocks) * block_size);
```

File: src/crimson/os/seastore/lba_manager/btree/btree_lba_manager.cc [hunk 2]
```diff
@@ -126,6 +126,7 @@ BtreeLBAManager::alloc_extent(
   extent_len_t len,
   paddr_t addr)
 {
+  ceph_assert((hint%(uint64_t)segment_manager.get_block_size()) == 0);
```
Comment (athanatos) [inline src/crimson/os/seastore/lba_manager/btree/btree_lba_manager.cc:129]: Do we have an is_aligned() helper?

Comment (liu-chunmei) [reply src/crimson/os/seastore/lba_manager/btree/btree_lba_manager.cc:129]: we have #define assert_aligned(x) ceph_assert(((x)%ctx.tm.get_block_size()) == 0) defined in object_data_handler.cc, but it uses ctx, (lba doesn't have this info),  I also can define another macro,   but if only one place use it, seems not necessary.  So what do you think?

Comment (athanatos) [reply src/crimson/os/seastore/lba_manager/btree/btree_lba_manager.cc:129]: Let's define a helper in seastore_types.h bool is_aligned(uint64_t offset, uint64_t alignment);

**Review by athanatos — APPROVED** [review]

**Pull request merged**